Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py

```diff
@@ -397,67 +354,128 @@ def _foregroundWidths(self):
+                OutputWorkspace=peakWSName,
+                RangeLower=xmin,
+                RangeUpper=xmax,
+                StartWorkspaceIndex=self.getProperty(Prop.START_WS_INDEX).value,
+                EndWorkspaceIndex=self.getProperty(Prop.END_WS_INDEX).value
+            )
+            linePosition = peak.LineCentre
+            self._cleanup.cleanup(peak.OutputWorkspace)
+        # Add the fractional workspace index of the beam position to the sample logs of ws.
+        ws.run().addProperty(common.SampleLogs.LINE_POSITION, float(linePosition), True)
+        return ws, linePosition
+
+    def _addSampleLogInfo(self, ws, linePosition):
+        """Add foreground indices (start, center, end), names start with reduction."""
+        run = ws.run()
+        hws = self._foregroundWidths()
+        beamPosIndex = int(numpy.rint(linePosition))
+        sign = self._workspaceIndexDirection(ws)
+        startIndex = beamPosIndex - sign * hws[0]
+        endIndex = beamPosIndex + sign * hws[1]
+        if startIndex > endIndex:
+            endIndex, startIndex = startIndex, endIndex
+        run.addProperty(common.SampleLogs.FOREGROUND_START, int(startIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_CENTRE, int(beamPosIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_END, int(endIndex), True)
+
+    def _moveDetector(self, ws, linePosition):
+        """Perform detector position correction for direct and reflected beams."""
+        detectorMovedWSName = self._names.withSuffix('detectors_moved')
+        twoTheta = ws.run().getProperty(common.SampleLogs.TWO_THETA).value
+        args = {
+            'InputWorkspace': ws,
+            'OutputWorkspace': detectorMovedWSName,
+            'TwoTheta': twoTheta,
+            'EnableLogging': self._subalgLogging,
+            'DetectorComponentName': 'detector',
+            'PixelSize': common.pixelSize(self._instrumentName),
+            'DetectorCorrectionType': 'RotateAroundSample',
+            'DetectorFacesSample': True
+        }
+        if not self.getProperty(Prop.TWO_THETA).isDefault:
+            # We should use user angle
+            args['TwoTheta'] = self.getProperty(Prop.TWO_THETA).value
+            # We need to subtract an offsetAngle from user given TwoTheta
+            args['LinePosition'] = linePosition
         else:
-            if not self.getProperty(Prop.BEAM_CENTRE).isDefault:
-                peakPos = self.getProperty(Prop.BEAM_CENTRE).value
-                beamPosWS = self._createFakePeakPositionTable(peakPos)
-            else:
-                # Beam position will be fitted later.
-                beamPosWS = None
-        return ws, beamPosWS
+            logs = ws.run()
+            args['TwoTheta'] = twoTheta + common.deflectionAngle(logs)
+        detectorMovedWS = SpecularReflectionPositionCorrect(**args)
+        self._cleanup.cleanup(ws)
+        return detectorMovedWS
+
+    def _reflectedBeamCalibration(self, ws):
```

Comment:
The name could be a bit misleading. Perhaps this function should be called something like `_calibrateDetectorAngleByDirect()`. The docstring below could be more clear about this, too.